File: solutions/fortran/allergies/1/allergies.f90

module allergies
   implicit none

   character(*), parameter :: allergiesPossible(*) = ["eggs", "peanuts", "shellfish", "strawberries", &
                                                      "tomatoes", "chocolate", "pollen", "cats"]
contains
   integer function toKey(allergic_str)
      character(*), intent(in) :: allergic_str
      integer :: position(1)
      position = findloc(allergiesPossible, allergic_str)
      toKey = lshift(1, position(1) - 1)
   end function

   logical function allergicTo(allergy_str, allergy_key)
      character(len=*), intent(in) :: allergy_str
      integer, intent(in) :: allergy_key
      allergicTo = iand(toKey(allergy_str), allergy_key) /= 0
   end function

   function allergicList(allergy_key)
      integer, intent(in) :: allergy_key
      character(len=100) :: allergicList
      integer :: i
      allergicList = ""
      do i = 1, size(allergiesPossible)
         if (allergicTo(allergiesPossible(i), allergy_key)) then
            allergicList = trim(allergicList)//" "//allergiesPossible(i)
         end if
      end do
      allergicList = trim(adjustl(allergicList))
   end function

end module

File: solutions/nim/allergies/1/allergies.nim
import std/tables
import std/setutils

type
  Allergen* = enum
    Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats

const allergens = {Eggs: 1, Peanuts: 2, Shellfish: 4, Strawberries: 8,
    Tomatoes: 16, Chocolate: 32, Pollen: 64, Cats: 128}.toTable

proc isAllergicTo*(score: int, allergen: Allergen): bool =
  (score and allergens[allergen]) > 0

func allergies*(score: int): set[Allergen] =
  var allergies: seq[Allergen] = @[]
  for allergen in allergens.keys:
    if isAllergicTo(score, allergen):
      allergies.add(allergen)
  return allergies.toSet
